File: lib/tomtom.R
library(fs)

ROOT <- path(find_root(has_file(".gitignore")))

source(path(ROOT, "lib", "common.R"))
source(path(ROOT, "lib", "load.R"))

################################### GLOBALS ####################################
CONFIG <- read_yaml(path(ROOT, "lib", "config.yaml"))

################################### PROCESSING #################################
compute_driving_df <- function(segs_df, beats_df, speed_exposure_df, prefix) {
  speed_buckets <-
    c(
      "below_neg_10",
      "neg_10_neg_5",
      "neg_5_to_0",
      "pos_0_to_5",
      "pos_5_to_10",
      "pos_10_to_15",
      "pos_15_to_20",
      "pos_20_to_25",
      "pos_25_to_30",
      "above_pos_30"
    )  
  segs_df %>%
    rename(
      lat_start = lat1, 
      lat_end = lat2, 
      lon_start = lon1, 
      lon_end = lon2
    ) %>%
    pivot_longer(
      cols = c(lat_start, lat_end, lon_start, lon_end),
      names_to = c('.value', "location"),
      names_sep = "_"
    ) %>%
    st_as_sf(coords = c("lon", "lat"), crs = 4269) %>%
    st_join(beats_df, join = st_within) %>%
    st_drop_geometry() %>% 
    drop_na(beat) %>%
    select(id, dist, beat) %>%
    group_by(id, beat) %>%
    # NOTE: Distance is split between the start and end beats of a segment.
    summarize(len = sum(dist / 2), .groups = "drop") %>%
    # Join on speeding + exposure data for each road
    left_join(speed_exposure_df, by = 'id') %>%
    # Within each beat, combine all roads of the same speed limit by:
    #   (1) Computing the average speeding behavior weighted by exposure,
    #   (2) Summing the total exposure of those roads.
    mutate(km_driven = tomtom_exposure * len) %>%
    select(-len, -tomtom_exposure, -id) %>%
    filter(km_driven != 0) %>%        
    mutate(across(speed_buckets, ~ . * km_driven)) %>%    
    group_by(beat, speed_limit) %>%
    summarize(across(everything(), sum)) %>%
    mutate(across(speed_buckets, ~ . / km_driven)) %>%
    # Match CMT dataframe specifications
    # NOTE: These fc are not real, they're imputed from the speed limit
    # NOTE: These are not speeding times: really, they're exposure.   
    mutate(
      fc = if_else(speed_limit > 72, 1, 2),
      beat = str_c(prefix, beat)
    )  %>%
    rename(speed_limit_kmh = speed_limit, geoid = beat) %>%
    pivot_longer(
      below_neg_10:above_pos_30,
      names_to = "sub_type",
      values_to = "speed_time"
    ) %>%
    mutate(speed_time = speed_time * km_driven) %>%
    mutate(
      sub_type = case_when(
        sub_type == 'neg_10_neg_5' ~ 'speeding -10 to -5',
        sub_type == 'neg_5_to_0' ~ 'speeding -5 to 0',
        sub_type == 'pos_0_to_5' ~ 'speeding 0 to 5',
        sub_type == 'pos_10_to_15' ~ 'speeding 10 to 15',
        sub_type == 'pos_15_to_20' ~ 'speeding 15 to 20',
        sub_type == 'pos_20_to_25' ~ 'speeding 20 to 25',
        sub_type == 'pos_25_to_30' ~ 'speeding 25 to 30',
        sub_type == 'above_pos_30' ~ 'speeding 30+',
        sub_type == 'pos_5_to_10' ~ 'speeding 5 to 10',
        sub_type == 'below_neg_10' ~ 'speeding below -10',
      )
    ) %>% 
    select(geoid, sub_type, fc, speed_limit_kmh, speed_time, km_driven)    
}

make_tomtom_driving_df <- function(city) {
  prefix <- pluck(CONFIG, city, 'geoid_prefix')
  beats_df <- load_beats(city) %>% clean_beats(city)
  segs_df <- read_csv(
    path(ROOT, 'data', city, 'segs', glue("{ city }_segs.csv"))
  )
  speed_exposure_df <- read_csv(
    path(ROOT, 'data', city, 'speeds', glue("{ city }_speed.csv"))
  )
  compute_driving_df(segs_df, beats_df, speed_exposure_df, prefix) %>%
    write_csv(
      path(ROOT, 'data', city, 'tomtom_driving', glue("{ city }_driving.csv"))
    )
}

File: lib/tomtom.py
#!/usr/bin/env python3
import os
import sys
import argparse
import json
import time
from collections import defaultdict

import git
import geopy.distance
import pandas as pd
import numpy as np

ROOT = git.Repo('.', search_parent_directories=True).working_tree_dir

def parse_road_segments(road):
    shapes = road['shape']
    n_points = len(shapes)
    data = []
    if n_points > 1:
        road_segments = []
        start = (shapes[0]['latitude'], shapes[0]['longitude'])        
        for p in shapes[1:]:
            end = (p['latitude'], p['longitude'])
            dist = geopy.distance.geodesic(start, end).m
            data.append(
                [road['segmentId'], start[0], start[1], end[0], end[1], dist]
            )
            start = end
    colnames = ['id', 'lat1', 'lon1', 'lat2', 'lon2', 'dist']
    return(pd.DataFrame(data = data, columns = colnames))

def parse_speeding(road):
    road_id = road['segmentId']
    road_seg_time_results = road['segmentTimeResults'][0]
    speed_limit = road['speedLimit'] 
    tomtom_exposure = road_seg_time_results['sampleSize']
    percentiles = road_seg_time_results['speedPercentiles']
    hist = np.histogram(
        a = [p - speed_limit for p in percentiles], 
        bins = np.r_[-np.inf, range(-10, 35, 5), np.inf], 
        density=False
    )
    row_vals = (list(hist[0]/len(percentiles)) 
                + [road_id, speed_limit, tomtom_exposure])
    colnames = [
        'below_neg_10','neg_10_neg_5', 'neg_5_to_0', 
        'pos_0_to_5', 'pos_5_to_10', 'pos_10_to_15', 
        'pos_15_to_20', 'pos_20_to_25', 'pos_25_to_30',
        'above_pos_30', 'id', 'speed_limit', 'tomtom_exposure'
    ]
    return(pd.DataFrame(data = [row_vals], columns = colnames))
    
def parse_tomtom(city):
    data_file = os.path.join(ROOT, 'data', city, 'json', f'{ city }.json')
    with open(data_file) as json_file:
        data = json.load(json_file)
    segment_rows = [parse_road_segments(road) for road 
                    in data['network']['segmentResults']]
    speed_rows = [parse_speeding(road) for road 
                  in data['network']['segmentResults']]
    segment_df = pd.concat(segment_rows)
    segment_df.to_csv(
        os.path.join(ROOT, 'data', city, 'segs', f'{ city }_segs.csv'), 
        index = False
    )
    speed_df = pd.concat(speed_rows)
    speed_df.to_csv(
        os.path.join(ROOT, 'data', city, 'speeds', f'{ city }_speed.csv'),
        index = False
    )

def parse_args(argv):
    desc = ('Parses json files from Tomtom into road segments '
            'and speeding distributions.'
           )
    parser = argparse.ArgumentParser(prog=argv[0], description=desc)
    parser.add_argument('cities', nargs='+')
    return parser.parse_args(argv[1:])

if __name__ == '__main__':
    args = parse_args(sys.argv)
    cities = args.cities
    for city in cities:
        parse_tomtom(city)
